### ISSUE TITLE
Consider update and accumulate methods on Property

### ISSUE THREAD

**Issue by kaifox** [ossgang/ossgang-commons#60]
In addition to pure set() methods, a property could have update-like methods like an Atomic reference which should then ensure atomic updates of the value. 

Proposed signatures:
```
Property<T>{
  update(UnaryOperator<T> updateOp);
  accumulate(T in, BinaryOperator<T> updateOp);

 // potentially also these variants (or only those)
  updateAndGet(UnaryOperator<T> updateOp);
  accumulateAndGet(T in, BinaryOperator<T> updateOp);
}
```

In principle, e.g. update would do the following in an atomic way:
```
T value = get();
set(updateOp.apply(value));
```

**Comment by andreacalia** [comment]
Ciao @kaifox ! Then it'll really become a "listenable" AtomicReference :smile:

**Comment by michi42** [comment]
Sounds fine to me :-)

We would probably have to extend `DispatchingObservableValue` a bit to get access to the internal `AtomicReference` it holds (we should not introduce another one in `Property` I think). But since this is our own internal class - I don't see any problem with that.

**Comment by kaifox** [comment]
Yes indeed, then it is a listenable atomic reference ;-) ... Indeed we should not have a second atomic reference ... Yes and the listener notification and the derived will not be atomic... but I think it still is worth it ...

I use such constructs a lot e.g. when you want to publish a state of e.g. a service and one has to do consistent modifications (transformations) of it. And in principle you want to make sure that you do not mix inputs, but 'chain' them. So at the moment one either has to synchronize or enforce a single thread ... both some effort if you use the pattern more frequently ... 

With an 'ObservableAtomicReference' it would be simpler ;-) ...

**Comment by michi42** [comment]
Yes, I agree, actually I always thought of Property as an "AtomicReference plus". So if you want to give it a try, please do.

**Comment by kaifox** [comment]
I will try (soon) ;-)

**Comment by kaifox** [comment]
Gi guys, I finally started trying this ... However, soon found a bit a showstopper:
With all these operations (accumulate, update, ... ) one either can get the old value *OR* the new value, but it is not possible to get both in a threadsafe consistent was. For the dispatching 'on change' we would need both ....

Currently, I have not idea how to solve this cleanly... Any proposals?

**Comment by michi42** [comment]
Perhaps we could use getAndUpdate() to fetch the previous value, and within the actual update callback we save the new value before we return?

**Comment by kaifox** [comment]
@michi42, do you mean more or less something like this?:
```java

public T getAndUpdate(UnaryOperator<T> operator) {
  AtomicReference<T> newValue = new AtomicReference<>();
  T oldValue = lastValue.getAndUpdate(x -> {
     T v = operator.apply(x);
     newValue.set(v);
     return v;
  })
  dispatch(newValue.get());
  return oldValue;
}

... It all looks a bit clumsy and I am completely unsure if it is bulletproof ....  I am not sure, if e.g. the update operations might be retried etc... 

```

**Comment by andreacalia** [comment]
Hi guys,

I didn't try it unfortunately.. but if `void DispatchingObservableValue.dispatchValue(T)` becomes something like `T DispatchingObservableValue.updateAndDispatchValue(UnaryOperator<T>)`? In the line of Michi's comment I think dispatch should happens atomically :thinking:

**Comment by michi42** [comment]
Hi,

From what I see, the function can be applied multiple times, but it is always the last result that will be propagated - so the hack you do should be safe as far as I can see (in the current implementation of AtomicReference that is...)
It is a bit a pity that AtomicReference has no getUpdateAndGet() method, since all the information is there in the implementations ...

**Comment by kaifox** [comment]
Ok ... I had similar thoughts ... so if you consider it 'ok' I will try to make a proposal based on this...

**Comment by kaifox** [comment]
@andreacalia , the problem is that - for the update on change notification - one needs both, old and final value ...

**Comment by andreacalia** [comment]
true indeed ! I see it now the problem.
The solution is a bit convoluted but should work as mentioned by michi :wink: .. anyway I don't see other solutions without introducing many more changes :innocent: 
@kaifox if you have a branch locally rebase first because there's an update that could affect you :slightly_smiling_face:

**Comment by kaifox** [comment]
There is a PR out #62 ... but I am not sure ... did I mess up something with the rebase? damn ...

**Comment by andreacalia** [comment]
Ciao Kajetan, I think it's ok just rebase on top of master now and it should be perfectly in sync :slightly_smiling_face:
